File: testsuite/Jogl4Examples/ShadersExample/src/OfflineControl.java
import java.io.File;

import model.ShaderOperationMode;
import model.ShadersModel;
import options.CommandLineOptions;
import render.OpenGlOfflineSphereRenderer;
import render.SoftwareRaycaster;
import vsdk.toolkit.common.linealAlgebra.Matrix4x4;
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.io.image.ImagePersistence;
import vsdk.toolkit.media.RGBImageUncompressed;

public final class OfflineControl
{
    private OfflineControl()
    {
    }

    public static void run(CommandLineOptions options)
    {
        ShadersModel model = ShadersModel.createDefault();
        model.updateSoftwareViewportAndCamera(options.getWidth(), options.getHeight());
        applyRenderFeatureOverrides(model, options);

        if ( options.getRotationDegrees() != null ) {
            model.setSphereRotationAngleRadians(
                Math.toRadians(options.getRotationDegrees()));
        }
        if ( options.getLightRotationDegrees() != null ) {
            applyOfflineLightRotation(
                model,
                Math.toRadians(options.getLightRotationDegrees().doubleValue()));
        }

        Matrix4x4 modelRotation = new Matrix4x4().axisRotation(
            model.getSphereRotationAngleRadians(),
            0.0,
            0.0,
            1.0);

        RGBImageUncompressed output;
        if ( options.getMethod() == ShaderOperationMode.SOFTWARE ) {
            SoftwareRaycaster raycaster = new SoftwareRaycaster();
            raycaster.render(model, model.getCamera(), modelRotation);
            output = model.getSoftwareFrameImage();
        }
        else {
            OpenGlOfflineSphereRenderer renderer = new OpenGlOfflineSphereRenderer();
            output = renderer.render(
                model,
                modelRotation,
                options.getWidth(),
                options.getHeight());
        }

        File outputFile = new File(options.getOfflineOutputPath());
        File parent = outputFile.getParentFile();
        if ( parent != null && !parent.exists() ) {
            parent.mkdirs();
        }
        ImagePersistence.exportPNG(outputFile, output);
        System.out.println("Offline render exported: " + outputFile.getAbsolutePath());
    }

    private static void applyOfflineLightRotation(
        ShadersModel model,
        double lightRotationRadians)
    {
        Matrix4x4 rotation = new Matrix4x4().axisRotation(
            lightRotationRadians,
            0.0,
            -1.0,
            0.0);
        Vector3D baseLightPosition = new Vector3D(1.0, -3.0, 1.0);
        model.getLight().setPosition(rotation.multiply(baseLightPosition));
    }

    private static void applyRenderFeatureOverrides(
        ShadersModel model,
        CommandLineOptions options)
    {
        if ( options.getWithTexture() != null ) {
            model.getQuality().setTexture(options.getWithTexture().booleanValue());
        }
        if ( options.getWithBumpMap() != null ) {
            model.getQuality().setBumpMap(options.getWithBumpMap().booleanValue());
        }
    }
}

File: cpp/testsuite/OpenGL4Examples/ShadersExample/src/OfflineControl.cpp
#include <cmath>
#include <cstdio>

#include "java/io/File.h"
#include "OfflineControl.h"
#include "model/ShadersModel.h"
#include "render/OpenGlOfflineSphereRenderer.h"
#include "render/SoftwareRaycaster.h"
#include "vsdk/toolkit/common/linealAlgebra/Matrix4x4d.h"
#include "vsdk/toolkit/common/linealAlgebra/Vector3Dd.h"
#include "options/CommandLineOptions.h"
#include "vsdk/toolkit/io/image/ImagePersistence.h"
#include "vsdk/toolkit/render/shaders/CpuTextureSamplingConfig.h"
#include "vsdk/toolkit/render/opengl4/OpenGL4ImageRenderer.h"
int OfflineControl::run(const CommandLineOptions& options)
{
    ShadersModel model = ShadersModel::createDefault();
    model.showHud = false;
    model.updateSoftwareViewportAndCamera(options.width, options.height);

    if (options.hasWithTexture) model.quality.setTexture(options.withTexture);
    if (options.hasWithBumpMap) model.quality.setBumpMap(options.withBumpMap);
    if (options.hasShadingType) model.quality.setShadingType(options.shadingType);
    if (options.hasMeridians) model.sphereMeridians = options.meridians;
    if (options.hasParallels) model.sphereParallels = options.parallels;
    if (options.hasTextureFilter) {
        OpenGL4ImageRenderer::setTextureFilterMode(
            options.textureFilter == CommandLineOptions::TextureFilterOption::NEAREST
                ? OpenGL4ImageRenderer::TextureFilterMode::NEAREST
                : OpenGL4ImageRenderer::TextureFilterMode::LINEAR);
    }
    if (options.hasCpuTextureOffsetU || options.hasCpuTextureOffsetV) {
        const double u = options.hasCpuTextureOffsetU ? options.cpuTextureOffsetUTexels : -0.5;
        const double v = options.hasCpuTextureOffsetV ? options.cpuTextureOffsetVTexels : -0.5;
        CpuTextureSamplingConfig::setTextureOffsetTexels(u, v);
    }
    if (options.hasRotation) model.setSphereRotationAngleRadians(options.rotationDegrees * M_PI / 180.0);
    if (options.hasLightRotation) {
        Matrix4x4d rot = Matrix4x4d().axisRotation(options.lightRotationDegrees * M_PI / 180.0, 0.0, -1.0, 0.0);
        model.light->setPosition(rot.multiply(Vector3Dd(1.0, -3.0, 1.0)));
    }

    Matrix4x4d modelRotation = Matrix4x4d().axisRotation(model.sphereRotationAngleRadians, 0.0, 0.0, 1.0);
    RGBImageUncompressed* output = 0;

    if (options.method == ShaderOperationMode::SOFTWARE) {
        SoftwareRaycaster raycaster;
        raycaster.render(&model, model.camera, modelRotation);
        output = model.softwareFrameImage ? model.softwareFrameImage->clone() : 0;
    }
    else {
        OpenGlOfflineSphereRenderer renderer;
        output = renderer.render(&model, modelRotation, options.width, options.height);
    }

    if (!output) return 1;
    java::File outputFile(options.offlineOutputPath.c_str());
    bool ok = ImagePersistence::exportPNG(outputFile, output);
    delete output;
    if (!ok) return 1;

    std::printf("Offline render exported: %s\n", options.offlineOutputPath.c_str());
    return 0;
}
